Add Tuya button support

dschall opened the pull request:
This change adds support for various Tuya / Zemismart battery-operated switches.
These devices send custom Zigbee commands (ID: 253) and expose duplicate PowerConfiguration endpoints.

dschall commented on org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:
```diff
@@ -0,0 +1,215 @@
+ * <p>
+ * As the configuration is done via channel properties, this converter is usable via static thing types only.
+ *
+ * @author Henning Sudbrock - initial contribution
+ * @author Thomas Weißschuh - support for attribute-based buttons
```
Update authors

dschall replied:
.. and comment

dschall commented on org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:
```diff
@@ -0,0 +1,215 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
```
Update year

dschall commented on org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/command/TuyaButtonPressCommand.java:
```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
```
Update year

dschall commented on org.openhab.binding.zigbee/src/main/resources/OH-INF/thing/tuya/ts0044.xml:
```diff
@@ -0,0 +1,76 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="tuya_ts0044" listed="false">	
```
I think the button mapping for the 4-button switch is incorrect, Thomas (@oggerschummer) can provide the correct mapping.

Oggerschummer replied:
Just checked with my second switch. The ordering you have is correct. The first switch shipped with an error in the manual where the buttons where numbered differently.
No change required.

dschall commented on org.openhab.binding.zigbee/src/main/resources/discovery.txt:
```diff
@@ -11,3 +11,7 @@ xiaomi_lumisensor-switchaq2,modelId=lumi.sensor_switch.aq2
 xiaomi_lumisensorwaterleak,modelId=lumi.sensor_wleak.aq1
 innr-rc-110,vendor=innr,modelId=RC 110
 osram-switch-4x-eu,vendor=OSRAM,modelId=Switch 4x EU-LIGHTIFY
+tuya_ts0041,vendor=_TZ3000_tk3s5tyg,modelId=TS0041
```
We could remove the vendor property or change this to a regex match, e.g. _TZ3000_\w{8} .

dschall replied:
removed the vendor instead. modelId should be sufficient.

dschall commented:
todo: add tests

cdjackson commented:
Thanks @dschall - at the very top level, I think the approach looks fine - thanks for looking at this. I've only quickly scanned the code to check a few things I had in mind that could be problematic and these look ok, so I will take a better look in a few days.

dschall commented:
> Thanks @dschall - at the very top level, I think the approach looks fine - thanks for looking at this. I've only quickly scanned the code to check a few things I had in mind that could be problematic and these look ok, so I will take a better look in a few days.

Hey @cdjackson, I've added tests and fixed a bug. I think now the PR is ready to merge.
I would appreciate comments.

dschall commented:
@cdjackson Hey Chris, can we merge this?

cdjackson commented:
Sorry - I have limited time right now and probably for about the next 2 weeks. Due to travel restrictions we have to move house in very short time.

dschall commented:
@cdjackson, gentle ping to please review this PR :)

suppressio commented:
Hi, what's the state of this project?

suppressio commented:
@dschall Hi... I'm interested too... another gentle ping to review this PR ...

cdjackson commented:
Sorry - I'm in the process of moving and have limited time. I will try and look at this in the next few weeks as things hopefully calm down. I will likely refactor this into a separate bundle.

dschall commented:
@cdjackson, gentle ping.
Are you going to rewrite this into a separate bundle, or should I resolve the conflicts and complete this PR (assuming you'd sign off on it).

gauthier12 commented:
Hi. I'm interested too. Maybe it is better to resolve the conflicts and merge instead of losing the development.
It is possible to refactor into a separate bundle later

deeja commented:
@cdjackson  - would you have any time to look at this again? I just got one of these switches and pretty excited to get it running! Will shout you a beer :D

cdjackson commented:
Sorry for the delay with this - I will try and take a look at this over the next few days.

dschall commented on org.openhab.binding.zigbee/src/main/resources/discovery.txt:
```diff
@@ -15,4 +15,7 @@ xiaomi_lumisensormagnet,modelId=lumi.sensor_magnet
 xiaomi_lumiremotemini,modelId=lumi.remote.b1acn01
 innr-rc-110,vendor=innr,modelId=RC 110
 osram-switch-4x-eu,vendor=OSRAM,modelId=Switch 4x EU-LIGHTIFY
-legrand_dimmer_without_neutral,vendor=\u001f Legrand,modelId=\u001f Dimmer switch w/o neutral
```
oh dang it, I think I accidentally removed that line :(

cdjackson commented on org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:
```diff
@@ -0,0 +1,216 @@
+           clientCluster.sendDefaultResponse(command, ZclStatus.SUCCESS);
+        }
+        else {
+           logger.warn("{} received unknown command {}", endpoint.getIeeeAddress(), command);
```
Being petty, but please addd a `:` after the `{}` for the address - it's good to keep things standard as sometimes processing of logs looks for certain features.

cdjackson commented on org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:
```diff
@@ -0,0 +1,216 @@
+
+    @Override
+    public boolean commandReceived(ZclCommand command) {
+        logger.debug("{} received command {}", endpoint.getIeeeAddress(), command);
```
Being petty, but please addd a `:` after the `{}` for the address - it's good to keep things standard as sometimes processing of logs looks for certain features.

dschall replied:
done, also, the others. I'll resolve without adding a separate comment to each.

cdjackson replied:
Thanks.

cdjackson commented on org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/converter/ZigBeeConverterTuyaButton.java:
```diff
@@ -0,0 +1,216 @@
+            case 2:
+                return CommonTriggerEvents.LONG_PRESSED;
+            default:
+                logger.warn("{} received unknown pressType {}", endpoint.getIeeeAddress(), pressType);
```
Being petty, but please addd a `:` after the `{}` for the address - it's good to keep things standard as sometimes processing of logs looks for certain features.

cdjackson commented:
@dschall firstly, sorry for the slight pettiness on the logging format but I like to try and keep things standard so any searches, or log viewer type processing has a consistent format to look for...

Otherwise I'm happy enough with this - it all looks good so if you can make those couple of small changes I'll merge.
